vulnerable_people_form/integrations/postcode_lookup_helper.py

```diff
@@ -30,50 +30,102 @@ def address_line_builder(lpi_info, slice):
+                                                      start_suffix=lpi_info.get('PAO_START_SUFFIX'),
+                                                      end_number=lpi_info.get('PAO_END_NUMBER'),
+                                                      end_suffix=lpi_info.get('PAO_END_SUFFIX'),
+                                                      text_value=lpi_info.get('PAO_TEXT'))
+    primary_property_identifier_is_a_building = True if lpi_info.get('PAO_TEXT') else False
+
+    secondary_property_identifier = property_identifier(start_number=lpi_info.get('SAO_START_NUMBER'),
+                                                        start_suffix=lpi_info.get('SAO_START_SUFFIX'),
+                                                        end_number=lpi_info.get('SAO_END_NUMBER'),
+                                                        end_suffix=lpi_info.get('SAO_END_SUFFIX'),
+                                                        text_value=lpi_info.get('SAO_TEXT'))
+
     street = lpi_info.get("STREET_DESCRIPTION", "")
+    organisation = lpi_info.get("ORGANISATION", "")
 
-    if sao != "":
-        building_and_street_line_1 = sao
-        building_and_street_line_2 = pao + " " + street
-    elif lpi_info.get("PAO_TEXT"):
-        building_and_street_line_1 = pao
-        building_and_street_line_2 = street
-    else:
-        building_and_street_line_1 = pao + " " + street
+    building_and_street_line_1, building_and_street_line_2 = build_building_and_street_lines(
+        primary_property_identifier, primary_property_identifier_is_a_building, secondary_property_identifier, street)
+
+    if organisation:
+        building_and_street_line_1, building_and_street_line_2 = include_organisation_name(building_and_street_line_1,
+                                                                                           building_and_street_line_2,
+                                                                                           organisation)
 
     return {
-        "uprn": int(lpi_info.get("UPRN")),
-        "town_city": address_line_builder(lpi_info, ["LOCALITY_NAME", "TOWN_NAME", "ADMINISTRATIVE_AREA"]).title(),
+        "uprn": int(lpi_info.get("UPRN")) if lpi_info.get("UPRN") else None,
+        "town_city": town_city_builder(lpi_info).title(),
         "postcode": lpi_info.get("POSTCODE_LOCATOR"),
         "building_and_street_line_1": building_and_street_line_1.title(),
         "building_and_street_line_2": building_and_street_line_2.title(),
     }
 
 
+def town_city_builder(lpi_info):
+    locality = lpi_info.get('LOCALITY_NAME')
+    town = lpi_info.get('TOWN_NAME')
+    administrative_area = lpi_info.get('ADMINISTRATIVE_AREA')
+
+    non_null_sections = [item for item in [locality, town, administrative_area] if item]
+
+    unique_sections = unique_values_preserving_order(non_null_sections)
+
+    return ', '.join(unique_sections)
+
+
+def unique_values_preserving_order(sections):
```

Comment:
I think this can just be list(dict.fromkeys(non_null_sections)), as dicts are now order preserving in python 3.6+

Reply:
Nice alternative but unless you consider this a blocking issue I prefer the verbose version

Reply:
Not blocking :)